### PR TITLE
[Video] Fix Audio Codec Description for DVD Playback

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

Aligned the information provided for stream selection of DVD audio streams with ffmpeg demux streams.
Use the proper fields and favor ffmpeg stream info (more reliable/detailed) when available. For example:

before:
name = (Director's comments) AC3 5.1
codecDesc = empty
bitrate = empty

after:
name = (Director's comments)
codecDesc = AC3 5.1(side)
bitrate = 448000

ffmpeg info provides a more accurate channels count and description (ex. with IFO nav data: 4 channels, with ffmpeg: 3.1 or quad)

Notes:
- Filling the bitrate field has an impact in the new stream selection dialogs since it's a factor in the streams sorting.
- Some info can be provided by both IFO and demuxer data. I didn't eliminate the partial duplication of the decoding because when a video starts (ie before and during menus, going to main feature and back), CSelectionStreams::Update() seems to be called twice. First with null ffmpeg demuxer (can't rely on ffmpeg info) and then a second time with a valid demuxer (prioritize the demux info). I don't know if the second call always happens and didn't want to disturb something that works, likely receives relatively little testing/use and maybe nobody in the team knows anymore.

Maybe in the future the name could be empty in favor of StreamFlags usage - visually impaired IFO attribute is handled throughout Kodi as a flag value but commentaries are not (assuming that's what StreamFlags::FLAG_COMMENT is for)

## Motivation and context

needed for #25947.

Currently when playing a DVD, an audio stream name contains codec name + channel layout and the codec description is empty. This makes things complicated for skins.

For DVD the codec description was built only from IFO information but there are sometimes inaccuracies (channels count in one of my samples) and it doesn't take into account the backwards compatible DTS-ES found on some DVD.

Channel counts with multiple layouts (ex. 4 channels can be 3.1, 4.0) show as 4 channels, even when more detailed info is available in the stream.

## How has this been tested?

A few DVD with DD+, ac3 stereo, dts streams.
Don't have a DVD with dts-es but tested with an dts-es in mkv that a correct codec desc is built.

Tested DVD with a DTS track declared as 5 channels in IFO > is actually 6 channels, now displayed as such.
1 channel volume is lower than the others, so it's a bit like 4.1 but technically it isn't.

Tested DVD with quad AC3 track. Showed 4 channels, now shows quad (side).

## What is the effect on users?

More consistent / accurate display of DVD audio streams information.

## Screenshots (if appropriate):

The change is not very significant at this point because the skins work around the problem but it will allow more interesting display of the information.
before

image

image

after (like other ffmpeg demuxer streams)

image

image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp [hunk 1]
```diff
@@ -1026,60 +1026,56 @@ void CDVDInputStreamNavigator::SetAudioStreamName(AudioStreamInfo &info, const a
...
     break;
   case 7:
-    info.name += " 6.1";
+    info.codecDesc += " 6.1";
     break;
   default:
-    char temp[32];
-    snprintf(temp, sizeof(temp), " %d-chs", audio_attributes.channels + 1);
-    info.name += temp;
+    info.codecDesc += StringUtils::Format(" {:d}-chs", audio_attributes.channels + 1);
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:1077]: use localized string
```suggestion
    info.codecDesc += StringUtils::Format(" {:d} {}", audio_attributes.channels + 1,
                                          g_localizeStrings.Get(10127));
```
and add the file path to the relative id on
addons\resource.language.en_gb\resources\strings.po

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 2]
```diff
@@ -439,10 +439,21 @@ void CSelectionStreams::Update(const std::shared_ptr<CDVDInputStream>& input,
...
     count = nav->GetAudioStreamCount();
     for(int i=0;i<count;i++)
     {
+      const auto stream =
+          std::find_if(demuxStreams.begin(), demuxStreams.end(),
+                       [i](const auto& stream)
+                       { return stream->type == STREAM_AUDIO && stream->dvdNavId == i; });
+      CDemuxStreamAudio* aStream =
+          (stream != demuxStreams.end()) ? static_cast<CDemuxStreamAudio*>(*stream) : nullptr;
+
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:456]: this can be full removed not really needed,
imo compare "dvdNavId" id to an index count could be not so good idea

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp (hunk 1 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:1077]: there is no space problem codecDesc is shown on the second line
immagine

**Review by ksooo — COMMENTED** [review]
I think those strings should be all loacalizable.

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp [hunk 3]
```diff
@@ -1026,60 +1026,56 @@ void CDVDInputStreamNavigator::SetAudioStreamName(AudioStreamInfo &info, const a
   switch(audio_attributes.audio_format)
   {
   case DVD_AUDIO_FORMAT_AC3:
-    info.name += " AC3";
+    info.codecDesc = "AC3";
     info.codecName = "ac3";
     break;
   case DVD_AUDIO_FORMAT_UNKNOWN_1:
-    info.name += " UNKNOWN #1";
+    info.codecDesc = "UNKNOWN #1";
```
[the remark this answers is not in the document]
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:1033]: i think maybe "unknown" cases can be grouped, end user know nothing about #1, #5, ....
“Other” case also to me doesn't make much sense

example:
info.codecName = "und"; ("und" iirc should exists for undetermined)
info.codecDesc = "unknown codec"

**Comment by Hitcher** [comment]
Posting here but I can open a new issue if you want.

`ListItem.Property(stream.codecdesc)` for PCM audio tracks in MKV's are shown as `0 channels`

screenshot00026

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp (hunk 1 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:1077]: if the problem the popup only
then should be found a way to improve the popup only without affect other parts
for example why the popup should have this big icon ? could be removed and save a lot of space for text
where the first line can show "Audio: english"
and second line the description

**Comment by jjd-uk** [comment]
> Posting here but I can open a new issue if you want.
> 
> `ListItem.Property(stream.codecdesc)` for PCM audio tracks in MKV's are shown as `0 channels`
> 
@CrystalP  see discussion in Slack **skin-estuary** channel, I've posted a sample there if you're willing to take a look.

**Comment by CrystalP** [comment]
Yes I'll localize (except the ones that are technical names and don't need translation).
I'm on the fence for MPEG AUDIO. To localize because of AUDIO or is it a standard name?

**Comment by CrystalP** [comment]
> Posting here but I can open a new issue if you want.
> 
> `ListItem.Property(stream.codecdesc)` for PCM audio tracks in MKV's are shown as `0 channels`

Sure, but out of scope for this PR, dedicated to DVD playback (because they have a separate code path).

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 2 above)
Comment (CrystalP) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:456]: The dvdNavId business is a little hard to follow but necessary to connect the ffmpeg audio streams list with the dvd nav audio stream info. I experienced mismatches when relying on stream enumeration order alone because the ffmpeg streams list gets shuffled occasionally and doesn't match the physical stream order anymore., while dvdnav info uses the physical order. VP has that dvdnavid to maintain the link between the two lists.

So, ordinal in dvdnav audio stream enumeration (i in the loop) = dvdnavid of the matching ffmpeg stream.
You can check other code in VP using dvdnavid and m_dvd. I can add a comment to explain some of this.

The IFO files contain a sample rate (48k or 96k), but not a bitrate from what I could see. They don't have a value defined for dts-es and sometimes an incorrect channel count, which is why I went looking for the info per ffmpeg. But it's a separate commit so it could be omitted if "the team" is not comfortable.
see http://www.mpucoder.com/DVD/ifo.html, "audio attributes" libdvdnav/read headers and code, dvd_types.h

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 4]
```diff
@@ -453,9 +464,18 @@ void CSelectionStreams::Update(const std::shared_ptr<CDVDInputStream>& input,
...
+      {
+        s.codecDesc = aStream->GetStreamType();
+        s.channels = aStream->iChannels;
+        s.bitrate = aStream->iBitRate;
+      }
+      else
+      {
+        s.codecDesc = info.codecDesc;
+        s.channels = info.channels;
+      }
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:477]: if you want keep this mixage you have to explain the reason because to me its unclear,
im running the original code without your "aStream" condition and i can get the codecDesc/channels correctly also on menu interludes

Comment (CrystalP) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:477]: It's not about the codecDesc in menus.
The reason is codec + channel count that seems more accurate from ffmpeg than in ifo: there is no value defined for DTS-ES in ifo, and I have a DVD where channel count in ifo is not accurate. Then, might as well use the opportunity to grab the bitrate,,not available in ifo. That one makes a small difference in the stream selection sort order.
I don't care that much about the commit, I'm not much of a DVD user. For the most part I just have a couple handy for regression testing.

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp [hunk 5]
```diff
@@ -1026,60 +1026,60 @@ void CDVDInputStreamNavigator::SetAudioStreamName(AudioStreamInfo &info, const a
...
+    CLog::LogF(LOGINFO, "unknown dvd audio codec DVD_AUDIO_FORMAT_UNKNOWN_1");
     break;
   case DVD_AUDIO_FORMAT_MPEG:
-    info.name += " MPEG AUDIO";
+    info.codecDesc = g_localizeStrings.Get(37003); // MPEG AUDIO
     info.codecName = "mp1";
     break;
   case DVD_AUDIO_FORMAT_MPEG2_EXT:
-    info.name += " MP2 Ext.";
+    info.codecDesc = "MP2 Ext.";
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:1041]: here can be also possible dont use translations just use full length name
"MPEG-1" for DVD_AUDIO_FORMAT_MPEG
"MPEG-2" for DVD_AUDIO_FORMAT_MPEG2_EXT

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp [hunk 6]
```diff
@@ -1026,60 +1026,60 @@ void CDVDInputStreamNavigator::SetAudioStreamName(AudioStreamInfo &info, const a
...
 
   switch(audio_attributes.channels + 1)
   {
   case 1:
-    info.name += " Mono";
+    info.codecDesc = g_localizeStrings.Get(37004); // mono
     break;
   case 2:
-    info.name += " Stereo";
+    info.codecDesc = g_localizeStrings.Get(37005); // stereo
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp:1071]: was "+=" with a prefix space

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 2 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:456]: i understood interesting link
so anyway i find a good thing add a little comment something like example
`// when possible prefer ffmpeg info because more reliable than IFO nav metadata`

**Comment by CrystalP** [comment]
Changes made.

**Review by CastagnaIT — APPROVED** [review]
from my part ok,
just add to string.po the file path for new use of id 10127
